### Name: Zheng CEN
Course: BUDT704 
Section: <0502> 
Date: <11/05/21> 

In [1]:
import numpy as np
import pandas as pd

import re

1. Consider the data within the Excel spreadsheet: 
a. (1 point) Load the data from the file into a data frame. 

In [2]:
df = pd.read_excel('HW5_Socially Responsible ETFs.xlsx')
df.head()

Ticker                        Fund Name     Issuer      AUM  Expense Ratio  \
0   ESGU   iShares ESG Aware MSCI USA ETF  Blackrock  $23.43B         0.0015   
1   ESGD  iShares ESG Aware MSCI EAFE ETF  Blackrock   $7.11B         0.0020   
2   ESGE    iShares ESG Aware MSCI EM ETF  Blackrock   $7.00B         0.0025   
3   ICLN  iShares Global Clean Energy ETF  Blackrock   $6.46B         0.0042   
4   ESGV      Vanguard ESG U.S. Stock ETF   Vanguard   $5.72B         0.0012   

  3-Mo TR                                            Segment  
0  0.0348                        Equity: U.S. - Total Market  
1  0.0165  Equity: Developed Markets Ex-North America - T...  
2  0.0147            Equity: Emerging Markets - Total Market  
3  0.0612                    Equity: Global Renewable Energy  
4  0.0355                        Equity: U.S. - Total Market

b. (1 point) Convert any missing values into NaN.
use pd.replace change '--' to NaN
and then show the position of them

In [3]:

df2 = df.replace('--','NaN')

#check the position of NAN
# print("each column has NaN\n", df2.isnull().sum(),"\n")
# print("total has NaN\n", df2.isnull().sum().sum(),"\n")

# print("which column values\n", df2.notnull(),"\n")


c. (1 point) Print a data frame containing the first 20 funds, including all data known about them. 
use pd.iloc function

In [4]:
# first 20rows
select_rows = 20
# use [0:20] just 1 to 20 lines
df2.iloc[:select_rows]

Ticker                                          Fund Name  \
0    ESGU                     iShares ESG Aware MSCI USA ETF   
1    ESGD                    iShares ESG Aware MSCI EAFE ETF   
2    ESGE                      iShares ESG Aware MSCI EM ETF   
3    ICLN                    iShares Global Clean Energy ETF   
4    ESGV                        Vanguard ESG U.S. Stock ETF   
5    XSOE  WisdomTree Emerging Markets ex-State-Owned Ent...   
6    SUSL                   iShares ESG MSCI USA Leaders ETF   
7    SUSA                    iShares MSCI USA ESG Select ETF   
8    USSG       Xtrackers MSCI U.S.A. ESG Leaders Equity ETF   
9     DSI                    iShares MSCI KLD 400 Social ETF   
10    TAN                                  Invesco Solar ETF   
11   QCLN  First Trust NASDAQ Clean Edge Green Energy Ind...   
12   VSGX               Vanguard ESG International Stock ETF   
13    PBW                Invesco WilderHill Clean Energy ETF   
14   EAGG          iShares ESG Aware U.S. Aggregate Bond ETF   
15   ESML           iShares ESG Aware MSCI USA Small-Cap ETF   
16   NULV                     Nuveen ESG Large-Cap Value ETF   
17   CRBN            iShares MSCI ACWI Low Carbon Target ETF   
18   NUSC                           Nuveen ESG Small-Cap ETF   
19   KRBN                      KraneShares Global Carbon ETF   

               Issuer      AUM  Expense Ratio 3-Mo TR  \
0           Blackrock  $23.43B         0.0015  0.0348   
1           Blackrock   $7.11B         0.0020  0.0165   
2           Blackrock   $7.00B         0.0025  0.0147   
3           Blackrock   $6.46B         0.0042  0.0612   
4            Vanguard   $5.72B         0.0012  0.0355   
5          WisdomTree   $4.28B         0.0032 -0.0013   
6           Blackrock   $4.18B         0.0010  0.0527   
7           Blackrock   $4.06B         0.0025  0.0439   
8       Deutsche Bank   $4.00B         0.0010  0.0498   
9           Blackrock   $3.75B         0.0025   0.054   
10            Invesco   $3.55B         0.0069  0.0902   
11        First Trust   $2.83B         0.0060  0.1126   
12           Vanguard   $2.82B         0.0017  0.0127   
13            Invesco   $1.89B         0.0061    0.03   
14          Blackrock   $1.63B         0.0010  -0.011   
15          Blackrock   $1.21B         0.0017  0.0437   
16  Nuveen Securities   $1.18B         0.0025  0.0343   
17          Blackrock   $1.17B         0.0020  0.0282   
18  Nuveen Securities   $1.11B         0.0030  0.0356   
19               CICC   $1.06B         0.0079  0.1732   

                                              Segment  
0                         Equity: U.S. - Total Market  
1   Equity: Developed Markets Ex-North America - T...  
2             Equity: Emerging Markets - Total Market  
3                     Equity: Global Renewable Energy  
4                         Equity: U.S. - Total Market  
5             Equity: Emerging Markets - Total Market  
6                         Equity: U.S. - Total Market  
7                         Equity: U.S. - Total Market  
8                         Equity: U.S. - Total Market  
9                         Equity: U.S. - Total Market  
10                    Equity: Global Renewable Energy  
11                      Equity: U.S. Renewable Energy  
12              Equity: Global Ex-U.S. - Total Market  
13                    Equity: Global Renewable Energy  
14  Fixed Income: U.S. - Broad Market, Broad-based...  
15                           Equity: U.S. - Small Cap  
16                     Equity: U.S. - Large Cap Value  
17                          Equity: Global Low Carbon  
18                           Equity: U.S. - Small Cap  
19                 Commodities: Energy Carbon Credits

2. (10 points) You want to analyze the assets under management (AUM); however, after reviewing the 
data you notice the units are not the same for each value. Determine a method to address this issue 
and then modify the data frame. 

use the map lambda here to format the string AUM, if the end is B then change it to M and multiple 1000 at the same time. 
So the Q1 should be named 'NaN' instead of NaN of be convenient.

In [5]:
df2['AUM'] = df2['AUM'].map(lambda x: '$'+ str(float(x[1:-1])*1000)+'M' if x[-1]=='B' else x) 


3. (10 points) The Segment variable contains more than just the segment. It contains the asset class (e.g., 
equity, fixed income, etc.), market (e.g., U.S., global, emerging markets, etc.), and the segment itself 
(e.g., Total Market, Mid Cap Growth, etc.). Process this data into three new columns in the data frame. 

The value without '-' is the main problem
I find them out and add the ' - '
market of missing part is U.S. Global developed  Energy. 4 items
use functions of  replace.    maintain

In [6]:
df3 = df2.copy()

In [7]:

# df3[['asset class', 'tmp1']]= df3['Segment'].str.split(':', expand = True) #split them into 2 first
# df3.drop(columns=['Segment'],inplace= True)

In [8]:
ser1 = df3['Segment'].str.contains(' - ')==False  # return ser of index do not have -
df3.loc[ser1, 'Segment'] = df3.loc[ser1, 'Segment'].str.replace('Global ','Global - ', regex=False)

df3.loc[ser1, 'Segment'] = df3.loc[ser1, 'Segment'].str.replace('U.S. ','U.S. - ', regex=False)
df3.loc[ser1, 'Segment'] = df3.loc[ser1, 'Segment'].str.replace('Energy ','Energy - ', regex=False)
df3.loc[ser1, 'Segment'] = df3.loc[ser1, 'Segment'].str.replace('Developed ','Developed - ', regex=False)


In [9]:
# to check
ser2 = df3['Segment'].str.contains(' - ')==False 
ser2

0      False
1      False
2      False
3      False
4      False
       ...  
161    False
162    False
163    False
164    False
165    False
Name: Segment, Length: 166, dtype: bool

In [10]:
df3[['asset class', 'tmp']] = df3['Segment'].str.split(': ', expand = True)  #tmp first

df3[['market', 'segment']] = df3['tmp'].str.split(' - ', expand = True)
df3.drop(columns= ['tmp', 'Segment'], inplace= True)
df3.head()

Ticker                        Fund Name     Issuer        AUM  \
0   ESGU   iShares ESG Aware MSCI USA ETF  Blackrock  $23430.0M   
1   ESGD  iShares ESG Aware MSCI EAFE ETF  Blackrock   $7110.0M   
2   ESGE    iShares ESG Aware MSCI EM ETF  Blackrock   $7000.0M   
3   ICLN  iShares Global Clean Energy ETF  Blackrock   $6460.0M   
4   ESGV      Vanguard ESG U.S. Stock ETF   Vanguard   $5720.0M   

   Expense Ratio 3-Mo TR asset class                              market  \
0         0.0015  0.0348      Equity                                U.S.   
1         0.0020  0.0165      Equity  Developed Markets Ex-North America   
2         0.0025  0.0147      Equity                    Emerging Markets   
3         0.0042  0.0612      Equity                              Global   
4         0.0012  0.0355      Equity                                U.S.   

            segment  
0      Total Market  
1      Total Market  
2      Total Market  
3  Renewable Energy  
4      Total Market

4. (5 points) Modify the data frame such that it is indexed in multiple levels - first by the issuer and then 
by the ticker. Sort the data frame first by issuer and then by ticker. Display the first fifty records. 

use set_index here with sort_index to sort and name the level

In [11]:
df4 = df2.copy()

df4 = df4.set_index(keys = ['Issuer','Ticker']).sort_index(level=[0,1])  
df4.head(50)

Fund Name  \
Issuer                       Ticker                                                      
Allianz                      EMNT         PIMCO Enhanced Short Maturity Active ESG ETF   
                             RAFE                              PIMCO RAFI ESG U.S. ETF   
Ameriprise Financial         ESGN    Columbia Sustainable International Equity Inco...   
                             ESGS          Columbia Sustainable U.S. Equity Income ETF   
Barclays                     GRN                             iPath Series B Carbon ETN   
                             RODI                    Barclays Return on Disability ETN   
                             WIL                      Barclays Women in Leadership ETN   
Beyond Investing             VEGN                               U.S. Vegan Climate ETF   
Blackrock                    BGRN                        iShares Global Green Bond ETF   
                             CRBN              iShares MSCI ACWI Low Carbon Target ETF   
                             DMXF                   iShares ESG Advanced MSCI EAFE ETF   
                             DSI                       iShares MSCI KLD 400 Social ETF   
                             EAGG            iShares ESG Aware U.S. Aggregate Bond ETF   
                             EAOA          iShares ESG Aware Aggressive Allocation ETF   
                             EAOK        iShares ESG Aware Conservative Allocation ETF   
                             EAOM            iShares ESG Aware Moderate Allocation ETF   
                             EAOR              iShares ESG Aware Growth Allocation ETF   
                             EMXF                     iShares ESG Advanced MSCI EM ETF   
                             ESGD                      iShares ESG Aware MSCI EAFE ETF   
                             ESGE                        iShares ESG Aware MSCI EM ETF   
                             ESGU                       iShares ESG Aware MSCI USA ETF   
                             ESML             iShares ESG Aware MSCI USA Small-Cap ETF   
                             EUSB       iShares ESG Advanced Total USD Bond Market ETF   
                             HYXF    iShares ESG Advanced High Yield Corporate Bond...   
                             ICLN                      iShares Global Clean Energy ETF   
                             LDEM                      iShares ESG MSCI EM Leaders ETF   
                             SDG                        iShares MSCI Global Impact ETF   
                             SUSA                      iShares MSCI USA ESG Select ETF   
                             SUSB    iShares ESG Aware 1-5 Year USD Corporate Bond ETF   
                             SUSC             iShares ESG Aware USD Corporate Bond ETF   
                             SUSL                     iShares ESG MSCI USA Leaders ETF   
                             USXF                    iShares ESG Advanced MSCI USA ETF   
                             XJH                  iShares ESG Screened S&P Mid-Cap ETF   
                             XJR                iShares ESG Screened S&P Small-Cap ETF   
                             XVV                      iShares ESG Screened S&P 500 ETF   
CICC                         KCCA          KraneShares California Carbon Allowance ETF   
                             KESG         KraneShares MSCI China ESG Leaders Index ETF   
                             KEUA            KraneShares European Carbon Allowance ETF   
                             KGRN    KraneShares MSCI China Clean Technology Index ETF   
                             KRBN                        KraneShares Global Carbon ETF   
Change Finance               CHGX    Change Finance U.S. Large Cap Fossil Fuel Free...   
Compagnie Lombard Odier SCmA ESUS        ETRACS 2x Leveraged MSCI USA ESG Focus TR ETN   
Crestview                    MDCP                    VictoryShares THB Mid Cap ESG ETF   
                           

5. (2 points) Create an observation: What kinds of social issues are supported by the various ETFs? How 
might you group them together to provide categories of social issues? 
 

6. (5 points) To determine where you might want to consider research, determine how many socially 
responsible ETFs are issued by each issuer? From this analysis, which issuer(s) do you recommend 
studying further? Why? 

count each issuer's fund by their appearence

In [12]:
df6 = df2.copy()
df6[['Issuer']].value_counts()

Issuer                                  
Blackrock                                   27
Nuveen Securities                           12
Deutsche Bank                               10
Mirae Asset Global Investments Co., Ltd.     9
Invesco                                      9
State Street Global Advisors                 8
Inspire                                      8
New York Life                                7
Northern Trust                               6
WisdomTree                                   6
Toroso Investments                           5
CICC                                         5
Franklin Templeton                           4
Barclays                                     3
First Trust                                  3
Vanguard                                     3
VanEck                                       3
Impact Shares                                3
Crestview                                    3
Allianz                                      2
Goldman Sachs      

First of all, the Blackrock should be further researched as it contain 27 funds much more than others

Nuveen Securities, Deutsche Bank, Mirae Asset Global Investments Co., Ltd., Invesco, State Street Global Advisors, Inspire     should also be further researched as they contain around 10 funds, pretty good

7. (5 points) Review the total assets under management for each issuer. From this analysis, which 
issuer(s) do you recommend studying further? Why? 

delete NaN first and then group and sum

In [13]:
df7 = df2.copy()
df7 = df7.drop(df7[df7['AUM']=='NaN'].index) #delete rows contain 'NaN'

df7[['Asset_int']] = df7['AUM'].map(lambda x: float(x[1:-1]) if x != 'NaN' else x) # add new line to float AUM
#df7['Asset_int'].map(lambda x: x=0 if x == 'NaN' else x)
df7['Asset_int'].groupby(df7['Issuer']).sum().sort_values(ascending= False)
#df7['Asset_int'].map(lambda x: df7['Asset_int'].groupby(df7['Issuer']).sum() if x != NaN else x)

Issuer
Blackrock                                   65469.72
Vanguard                                     8783.04
Invesco                                      6268.55
WisdomTree                                   5465.83
Deutsche Bank                                5417.78
Nuveen Securities                            4588.68
First Trust                                  3222.99
Mirae Asset Global Investments Co., Ltd.     1600.00
CICC                                         1296.33
Inspire                                      1128.98
SS&C                                         1010.00
New York Life                                 738.13
Northern Trust                                481.34
Franklin Templeton                            432.51
Goldman Sachs                                 349.68
Allianz                                       199.33
ETFMG                                         191.42
VanEck                                        178.40
Toroso Investments                     

still choose the Blackrock, cause it contain total assets of 65469.72, numch hhigher than others

SS&C, Inspire, CICC, Mirae Asset Global Investments Co., Ltd., First Trust, Nuveen Securities, Deutsche Bank, WisdomTree, Invesco, Vanguard   
are also pretty well according to the rank of total assets

8. (5 points) Review the assets under management for each fund. From this analysis, which issuer(s) do 
you recommend studying further? Why? 

In [14]:
# continue use the df of Q7
df7['Asset_int'].groupby(df7['Issuer']).mean().sort_values(ascending= False)



Issuer
Vanguard                                    2927.680000
Blackrock                                   2424.804444
First Trust                                 1074.330000
SS&C                                        1010.000000
WisdomTree                                   910.971667
Invesco                                      895.507143
Deutsche Bank                                541.778000
Nuveen Securities                            382.390000
CICC                                         259.266000
ETFMG                                        191.420000
Mirae Asset Global Investments Co., Ltd.     177.777778
Goldman Sachs                                174.840000
Wahed                                        155.100000
Inspire                                      141.122500
Nationwide                                   119.100000
Franklin Templeton                           108.127500
EMPIRICAL FINANCE LLC                        105.900000
New York Life                            

Vanguard this time! as the average amount of asset on each fund is highest
Blackrock is also good enough (much higher than other funds)
So it seem that the funds of Vanguard maintain higher quality.

9. (5 points) An expense ratio describes how much of a fund's assets are used for operating expenses, 
which will likely get passed onto investors; thus, a lower expense ratio is better as an investment. 
Analyze the expense ratios for the funds of each issuer. From this analysis, which issuer(s) do you 
recommend studying further? Why? 

In [15]:
df9 = df2.copy()
df9 = df9.drop(df9[df9['Expense Ratio']=='NaN'].index) #delete rows contain 'NaN'
df7['Expense Ratio'].groupby(df7['Issuer']).mean().sort_values(ascending= True)




Issuer
Humankind USA LLC                           0.001100
Vanguard                                    0.001367
JPMorgan Chase                              0.001500
Deutsche Bank                               0.001690
Blackrock                                   0.001900
Northern Trust                              0.002167
Allianz                                     0.002850
Nuveen Securities                           0.002992
Nationwide                                  0.003400
New York Life                               0.003471
WisdomTree                                  0.003533
Fidelity                                    0.003900
The Hartford                                0.003900
Ameriprise Financial                        0.004000
Tortoise                                    0.004000
VanEck                                      0.004133
Reflection Asset Management, LLC            0.004500
Mirae Asset Global Investments Co., Ltd.    0.004633
Goldman Sachs                          

Humankind USA LLC is rank 1 for the average expense ratio of each fund
which means it has the lowest ration on each fund, goo job!

10. (10 points) Analyze frequency distributions of the asset classes, markets, and segments for the funds of 
each issuer. From this analysis, make a recommendation on the type of fund your company should 
pursue, factoring in an asset class, market, and segment. What do you recommend? Why? 

In [16]:
df10= df3.copy()


In [17]:

df10['asset class'].groupby(df10['asset class']).size().sort_values(ascending= False)



asset class
Equity              133
Fixed Income         21
Asset Allocation      6
Commodities           4
Leveraged Equity      2
Name: asset class, dtype: int64

In [18]:
df10['market'].groupby(df10['market']).size().sort_values(ascending= False)



market
U.S.                                  92
Global                                38
Emerging Markets                      10
Developed Markets Ex-North America     6
Developed Markets Ex-U.S.              4
Energy                                 4
China                                  3
Developed Markets                      3
Global Ex-U.S.                         2
Developed                              1
Emerging Markets Low Carbon            1
India                                  1
North America                          1
Name: market, dtype: int64

In [19]:
df10['segment'].groupby(df10['segment']).size().sort_values(ascending= False)



segment
Total Market                                53
Large Cap                                   30
Renewable Energy                            17
Corporate, Broad-based Investment Grade      6
Environment                                  6
Small Cap                                    5
Carbon Credits                               4
Target Risk                                  4
Mid Cap                                      4
Low Carbon                                   4
Broad Market, Broad-based                    3
Broad Market, Broad-based Investment Gr…     3
Corporate, Broad-based High Yield            3
Large Cap Growth                             3
Target Outcome                               2
Large Cap Value                              2
Corporate, Broad-based Investment Grade…     2
Broad Market, Broad-based Investment …       2
High Dividend Yield                          1
Markets Water                                1
Government, Non-Native Curr…                 1
Mid C

Equity  US Total market is the highest frequence, so recommanded

11. (5 points) Another option to determine what the company should pursue is to look at the 3-month 
total return percentages. Determine the mean 3-month total return percentages factoring in a 
combination of asset class, market, and segment together. From this analysis, make a recommendation 
on the type of fund your company should pursue, factoring in an asset class, market, and segment. 
What do you recommend? Why?  

In [20]:
#astype(float)
#df10 = df10.replace('NaN',NaN)
df10['3-Mo TR']=df10['3-Mo TR'].astype(float)
df10.head()

Ticker                        Fund Name     Issuer        AUM  \
0   ESGU   iShares ESG Aware MSCI USA ETF  Blackrock  $23430.0M   
1   ESGD  iShares ESG Aware MSCI EAFE ETF  Blackrock   $7110.0M   
2   ESGE    iShares ESG Aware MSCI EM ETF  Blackrock   $7000.0M   
3   ICLN  iShares Global Clean Energy ETF  Blackrock   $6460.0M   
4   ESGV      Vanguard ESG U.S. Stock ETF   Vanguard   $5720.0M   

   Expense Ratio  3-Mo TR asset class                              market  \
0         0.0015   0.0348      Equity                                U.S.   
1         0.0020   0.0165      Equity  Developed Markets Ex-North America   
2         0.0025   0.0147      Equity                    Emerging Markets   
3         0.0042   0.0612      Equity                              Global   
4         0.0012   0.0355      Equity                                U.S.   

            segment  
0      Total Market  
1      Total Market  
2      Total Market  
3  Renewable Energy  
4      Total Market

In [21]:
df10.groupby(['asset class','market','segment']).mean().sort_values(by=['3-Mo TR'],ascending = False)


Expense Ratio  \
asset class      market                             segment                                                   
Commodities      Energy                             Carbon Credits                                 0.007800   
Equity           India                              Total Market                                   0.005800   
                 U.S.                               Renewable Energy                               0.005450   
                                                    Mid Cap Value                                  0.003000   
                 North America                      Total Market                                   0.005900   
                 Global                             Renewable Energy                               0.005967   
                 U.S.                               Large Cap Growth                               0.004450   
                                                    Mid Cap                                        0.004175   
                                                    Large Cap                                      0.004108   
                 Global                             REITs                                          0.006900   
                 U.S.                               Extended Market                                0.005700   
                                                    Small Cap                                      0.002520   
                                                    Low Carbon                                     0.001500   
                 Global                             Large Cap Growth                               0.005900   
                                                    Environment                                    0.005317   
                 U.S.                               Mid Cap Growth                                 0.003000   
                 Global                             Large Cap                                      0.005200   
                 U.S.                               Total Market                                   0.003496   
                                                    Large Cap Value                                0.003700   
Asset Allocation Global                             Target Outcome                                 0.011250   
Equity           China                              Total Market                                   0.005667   
                 Global                             Low Carbon                                     0.003500   
                 Developed                          Markets Water                                  0.004000   
                 Global                             Total Market                                   0.006133   
                 Developed Markets                  Total Market                                   0.002850   
                 Global Ex-U.S.                     Large Cap                                      0.008000   
                 Developed Markets Ex-U.S.          Large Cap                                      0.003633   
                 Global                             Water                                          0.005000   
                                                    Total Market Growth                            0.005300   
                 Developed Markets Ex-North America Total Market                                   0.002100   
                 Global Ex-U.S.                     Total Market                                   0.001700   
Asset Allocation U.S.                               Target Risk                                    0.001800   
Equity           Emerging Markets                   Total Market                                   0.002833   
                 Global                             Consumer                                       0.004500   
Fixed Income     U.S.                               Corporate, Broad-based High Yield             

Commodities	Energy	Carbon Credits	  is recomanded as the highest 3Mon TR

12. (5 points) A senior executive at your company has said holding socially responsible investments is not 
profitable compared to holding investments that do not address socially responsible causes. 
Incorporate the analysis you have done in this assignment (and/or additional analysis if you would like) 
with a small amount of Internet research on current investments to prove or disprove the executive. 

In [22]:

df10['3-Mo TR'].mean()


0.03134893617021276

Search from Internet, the investment are not socially responsible 3 Mon total percentage rate is about 1.7%
So there is no evidence that the socially responsible is not as profit as investment that not responsible

"I pledge on my honor that I have not given nor received any unauthorized 
assistance on this assignment." 
--zhengcen